### PR TITLE
refactor(mito2): implement RegionManifestManager

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Initial implementation of `RegionManifestManager`. Provides basic construct/recover and query API

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/mito2/src/manifest/storage.rs [hunk 1]
```diff
@@ -537,6 +516,27 @@ impl ManifestObjectStore {
     }
 }
 
+#[derive(Serialize, Deserialize, Debug)]
+struct CheckpointMetadata {
+    pub size: usize,
+    /// The latest version this checkpoint contains.
+    pub version: ManifestVersion,
+    pub checksum: Option<String>,
+    pub extend_metadata: Option<HashMap<String, String>>,
```
Comment (v0y4g3r) [inline src/mito2/src/manifest/storage.rs:525]: Do we need to wrap HashMap inside an Option?

**Review by v0y4g3r — APPROVED** [review]
LGTM

File: src/mito2/src/manifest/action.rs [hunk 2]
```diff
@@ -36,50 +51,39 @@ pub struct RegionChange {
...
 
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
-pub struct RegionRemove {
-    pub region_id: RegionId,
-}
-
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
 pub struct RegionEdit {
     pub region_version: VersionNumber,
-    pub flushed_sequence: Option<SequenceNumber>,
```
Comment (evenyag) [inline src/mito2/src/manifest/action.rs:47]: I think we still need to store the flushed sequence in the edit.

File: src/mito2/src/manifest/storage.rs [hunk 3]
```diff
@@ -537,6 +516,27 @@ impl ManifestObjectStore {
...
+}
+
+impl CheckpointMetadata {
+    fn encode(&self) -> Result<impl AsRef<[u8]>> {
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:529]:
```suggestion
    fn encode(&self) -> Result<Vec<u8>> {
```

File: src/mito2/src/manifest/action.rs [hunk 4]
```diff
@@ -114,16 +115,22 @@ impl RegionManifestDataBuilder {
...
-            committed_sequence: self.committed_sequence,
+            version: self.version.context(RegionVersionNotFoundSnafu)?,
         })
     }
 }
 
-// The checkpoint of region manifest, generated by checkpoint.
+/// The region version checkpoint
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct RegionVersion {
```
Comment (evenyag) [inline src/mito2/src/manifest/action.rs:128]: Could we rename it to `RegionVersionCheckpoint`, `VersionCheckpoint`, or `ManifestCheckpoint`? `RegionVersion` is easy to confuse with the existing `struct Version` under the `region` mod.

Comment (waynexia) [reply src/mito2/src/manifest/action.rs:133]: I lost in those version/sequence/number. Need a mermaid to cure :sob:

**Review by evenyag — APPROVED** [review]
The manifest is so messy...We need to refactor it after merging it into mito2...@waynexia

**Pull request merged**